### ISSUE TITLE
Negative usteps in DOSXYZ isource=20 with IAEA phsp source through shared library geometry

### ISSUE THREAD

**Issue by ojalaj** [nrc-cnrc/EGSnrc#269]
Hello.

I'm using current `fix-misc_CM_fixes` branch with changes included in Fix #186 as well.

I'm experiencing negative ustep errors (+1000 always --> stopping the simulations) with DOSXYZnrc isource=20 runs using IAEA phsp source through shared library geometry. I don't know is this related to issues #187 or #90, but for some reason the errors are generated when using similar Varian-generated phsp source that I assume @crcrewso is using in #187 with header info:
```
$RECORD_CONTENTS:
    1     // X is stored ?
    1     // Y is stored ?
    0     // Z is stored ?
    1     // U is stored ?
    1     // V is stored ?
    1     // W is stored ?
    1     // Weight is stored ? <--- Changed from 0 to 1 by addphsp
    0     // Extra floats stored ?
    1     // Extra longs stored ? <--- Changed from 0 to 1 by addphsp
    2     // LATCH EGS variable stored in the extralong array [ 0] <--- Added by addphsp

$RECORD_CONSTANT:
    26.7000     // Constant Z
    1.0         // Constant Weight <-- Manually added by me according to info on Varian webpage

$RECORD_LENGTH:
29   (21 in the original Varian phsp file)
```
One thing I suspect is the number of particles in the large phsp file, which exceeds 2^31-1 (the number is 2367622510). Could this be the reason for negative usteps? However addphsp (version 2016) concatenated this large file without problems and no warnings or errors are generated anywhere else.

What is also weird is that I have run BEAMnrc ISOURC=21 simulations with the phsp file above without any problems. 

One discrepancy I found is that the addphsp for the latest release prints `Warning: IAEA format phsp file does not store LATCH` when concatenating IAEA phsp files, but the manual on p. 107 says "In the case of BEAMnrc, two extra long integers, LATCH and the number of primary histories between this particle and the last particle scored, are always stored."

I have another IAEA phsp file for other Varian linac model with following header information:
```
$RECORD_CONTENTS:
    1     // X is stored ?
    1     // Y is stored ?
    1     // Z is stored ?
    1     // U is stored ?
    1     // V is stored ?
    1     // W is stored ?
    1     // Weight is stored ?
    1     // Extra floats stored ?
    2     // Extra longs stored ?
    3     // ZLAST variable stored in the extrafloat array [ 0]
    1     // Incremental history number stored in the extralong array [ 0]
    2     // LATCH EGS variable stored in the extralong array [ 1]

$RECORD_CONSTANT:

$RECORD_LENGTH:
41
```
with 2080480392 particles. The Z is anyhow constant value for all particles. Simulations with the file run fine.

While reading the current BEAMnrc manual I can't find out which `RECORD_CONTENTS` are really required and which ones are discarded even though they would be included in the header.

**Comment by ojalaj** [comment]
I've made some further debugging and the problem seems also to be with pure BEAMnrc ISOURC=21 simulations with the abovementioned phsp file, but it only happens, when for SYNCJAWS there are more than one different jaws positions. It doesn't matter whether it is step&shoot or dynamic. Please find an extract of my .egsinp file:
.
.
*********** start of CM SYNCJAWS with identifier Jaws  ***********
30, RMAX
Jaws
2, 1, # PAIRED BARS OR JAWS, field type
Y
X
path for sequence file
0, 0, 0, 0,
0, 0, 0, 4,
material
0, 0, 0, 5,
material
*********** start of CM SLABS with identifier AirGap2  ***********
30, RMAX
Second Slab
1, NSLABS
44.9, ZMIN
0.1, 0, 0, 0, 0, 0
AIR521ICRU
.
.
With a sequence file below, the simulation runs with no errors:
Sequence
2
0
27.74455, 35.49393, 2.91318, 3.72686, -2.77446, -3.54939,
35.84613, 43.61093, 0.07169, 0.08722, -4.09721, -4.98473,
1
27.74455, 35.49393, 2.91318, 3.72686, -2.77446, -3.54939,
35.84613, 43.61093, 0.07169, 0.08722, -4.09721, -4.98473,

But already with slightly different file there are -usteps, not to mention the sequence files larger number of control points:
Sequence
2
0
27.74455, 35.49393, 2.91318, 3.72686, -2.77446, -3.54939,
35.84613, 43.61093, 0.07169, 0.08722, -4.09721, -4.98473,
1
27.73718, 35.48450, 2.77372, 3.54845, -3.05109, -3.90330,
35.84613, 43.62105, 1.18292, 1.43949, -2.93221, -3.56820,

No other errors/warnings are printed. If I interpret the -ustep warning correctly, it takes place at Z ~43.6, at the bottom of SYNCJAWS.

**Comment by blakewalters** [comment]
Hi Jarkko:

Would you be able to send me the complete .egsinp file?  In the meantime, I'm going to take a closer look at how addphsp deals with the Varian Truebeam files.

Thanks,
Blake

**Comment by crcrewso** [comment]
@blakewalters I didn't notice any issues with the addition when I summed all of the addphsp files. Do you have 250 GB of free space? If so I could send them (somehow)

**Comment by blakewalters** [comment]
@crewso, where did you obtain the original files?  I'm assuming not from the IAEA phase space database.  I'm thinking I might be able to reproduce the error just with just a couple of them concatenated together.

**Comment by ojalaj** [comment]
I've sent the .egsinp file by email.

**Comment by crcrewso** [comment]
@blakewalters I have a myvarian account, I downloaded the newest V2 files straight from the website. 

When concatenating there were times when the total counts broke. I got past it by redownloading any file where things started to break. their Download process is not the best

**Comment by ojalaj** [comment]
Hmm, at least for me there have been no problems with downloading (from MyVarian).

**Comment by ojalaj** [comment]
Hi @blakewalters - have you been able to re-produce the issue? May I help you any further on this?

**Comment by blakewalters** [comment]
I haven't had a chance to try reproduce this yet.  I'm wondering if there's a way that @ojalaj or @crcrewso could send a couple of the Truebeam files to me.  Perhaps via Dropbox?

**Comment by crcrewso** [comment]
@blakewalters I'm going to read my MyVarian license agreement. If its acceptable by the usage agreement how should we touch base?

**Comment by blakewalters** [comment]
Okay @crcrewso.  How large are each of these files?  Depending on the size, it may be possible to share them in a Dropbox.  Otherwise, I believe there are other large file transfer services out there.

**Comment by rtownson** [comment]
It's pretty straightforward to get an account on MyVarian (at least it was several years ago), you could sign up for one Blake.

**Comment by marenaud** [comment]
Be careful though, the license agreement that Varian makes you sign to have access to the phase space files is pretty restrictive and you give up a lot of rights to the work you do.

**Comment by blakewalters** [comment]
Let me try that first.  For some reason I got the impression I have to be associated with a hospital/clinic.

**Comment by marenaud** [comment]
In case this helps the conversation, on our cluster we have the Varian phsp files that we concatenated manually (just by doing cat phsp_1 phsp_2 phsp_3 > total_phsp) and with the IAEAheader created by simply parsing the individual phsp headers and adding up the particle counts. We **don't** get any negative ustep errors when using source 20 going through a shared library which includes SYNCJAWS with many jaw positions. Our header looks like:
```
$RECORD_CONTENTS:
    1     // X is stored ?
    1     // Y is stored ?
    0     // Z is stored ?
    1     // U is stored ?
    1     // V is stored ?
    1     // W is stored ?  //W is not stored; set to 1 to enable BEAMnrc import
    0     // Weight is stored ?
    0     // Extra floats stored ?
    0     // Extra longs stored ?

$RECORD_CONSTANT:
     26.7       // Constant Z
     1.0000     // Constant Weight

$RECORD_LENGTH:
21
```

**Comment by luongoanh** [comment]
@ojalaj . Hi.Can you tell me how to get * .sequence inp file into SYNCJAWS (SYNCMLC ...). Thanks

**Comment by ftessier** [comment]
Fixed in pull request #278.
